### ISSUE TITLE
Hash for fraction field element is too permissive

### ISSUE THREAD

**Issue by xcaruso** [sagemath/sage#35238]
### Steps To Reproduce

```sage
sage: A.<x,y> = QQ[]
sage: K = A.fraction_field()
sage: d = { x/y: 0 }
sage: d[K(2*x, 2*y)]
Traceback (most recent call last):
...
KeyError: 2*x/(2*y)
```

### Expected Behavior

`(2*x)/(2*y)` is equal to `x/y`, so `d[K(2*x, 2*y)]` should return `0`.

### Actual Behavior

The reason why it fails is that `x/y` and `(2*x)/(2*y)` differ in their hash though they are considered as equal by equality test.

**Comment by tehami02** [comment]
I have worked on this problem and I think this could probably be the reason -
The reason why d[K(2*x, 2*y)] does not return 0 is because the derivative of K(2*x, 2*y) with respect to x and y is not zero.

While 2*x/(2*y) simplifies to x/y, their derivatives with respect to x and y are not the same. Specifically, the derivative of 2*x/(2*y) with respect to x is 2/(2*y), which simplifies to 1/y. On the other hand, the derivative of x/y with respect to x is 1/y, which is the same.

But when taking the partial derivative of K(2*x, 2*y) with respect to x, we get 2*y*exp(2*x*y), which is not zero. Similarly, when taking the partial derivative of K(2*x, 2*y) with respect to y, we get 2*x*exp(2*x*y), which is also not zero.

In summary, just because two expressions are algebraically equivalent does not mean that their derivatives will be the same.

**Comment by xcaruso** [comment]
Sorry, I don't understand your answer; how do you get these weird expression 2y*exp(2xy) and 2x*exp(2xy)?

**Comment by videlec** [comment]
Do you mean `K.<x, y>` in the first line?

**Comment by videlec** [comment]
To my mind, the only reasonable default for fraction field element hash is to always return `0`. But then, it does not make much sense to use these in a `set` or as keys of a `dict`. If you have additional structure on the ring (eg Euclidean domain, PID, grading) then smarter hashes could be designed.

**Comment by xcaruso** [comment]
Sorry, I don't understand your answer; how do you get these weird expression 2y*exp(2xy) and 2x*exp(2xy)?

> @xcaruso You did not defined `K`. I guess you forgot `K = A.fraction_field()`...

Yes, thanks! I edited my message.

**Comment by saraedum** [comment]
See also the report in https://groups.google.com/g/sage-support/c/H7A890hlrN8/m/8g58npoLDgAJ

**Comment by saraedum** [comment]
@videlec the only downside with just returning a constant 0 (apart from a performance hit) is that the docstring states:

```
        This function hashes in a special way to ensure that generators of
        a ring `R` and generators of a fraction field of `R` have the same
        hash. This enables them to be used as keys interchangeably in a
        dictionary (since ``==`` will claim them equal).
```

I don't really think that this is a terribly important feature to preserve.

**Comment by saraedum** [comment]
I guess we could just try to convert the element to the base ring and if that works take the hash there. If it fails but the element is in the base ring, then the hash is wrong but that's maybe an acceptable middle ground here.

Something like:

```py
if self._denominator.is_one():
    return hash(self._numerator)
if self._parent.is_exact():
    self.reduce()
try:
    return hash(self.base()(self))
except ...:
    pass
return 42    
```

**Comment by nbruin** [comment]
While I agree that replacing a malfunctioning hash with a functioning one isn't really damaging performance (because it was buggy before), I do think this deserves a better solution. Often denominators can be normalized. For fraction fields of (multivariate) function fields:
 - if the base ring is a field we can make the denominator monic
 - if the base ring in ZZ we can remove all content from the denominator and make the leading coefficient positive
More generally, in PIDs where we have preferred generators for ideals (so: orbit representatives for the action of the unit group) we can normalize the denominator; first by removing content and then picking a preferred leading coefficient.
Then we can just compute hashes of normalized numerators and denominators (and normalize fractions more fully)

**Comment by videlec** [comment]
@nbruin I agree that it could be made better in some cases. But these could be considered as two disjoint tasks
- fix the generic hash of `FractionFieldElement`
- implement useful hashes for derivative classes such as fraction field of polynomials

**Comment by nbruin** [comment]
@videlec That's one plan, but it would leave us with an atrocious generic hash implementation; so bad that perhaps it shouldn't be there? How bad would it be to make FractionFieldElements by default not hashable and then implement useful hashes for inheriting classes? I see that as an alternative way to address the first bullet? I know we sometimes need hashes because stuff gets put in dictionary keys. But that is part of the reason why I'm worried about providing a performance-breaking hash hack in. I think it's worth seeing which is worse: a bad hash or no hash at all. Do we have examples where we threw our hands up and went with a constant hash and not ended up in deeper trouble down the road?

**Comment by saraedum** [comment]
I would like to get #40000 in because I think it's better to produce the correct result slowly than the wrong result quickly.

I won't put any more work into this because properly fixing this is going to take several weeks realistically and I don't use fraction fields at all myself atm.

Obviously, it would be great to implement this properly and so if somebody wants to expand upon #40000 please do. (But imho it should happen as a followup because otherwise I am worried that this bug is going to stay in SageMath for a few more years…)

**Comment by videlec** [comment]
@nbruin Making `FractionFieldElement.__hash__` raise an error would indeed be an alternative to return a (not exactly) constant hash as proposed in #40000. However I am afraid that many doctests and user code contains dictionaries with fraction field keys. Will you advocate for a 1 year deprecation period while we keep the buggy hash and display a warning that claims that it will be forbidden to use fraction field element in dictionaries?

To my mind, any programmer relying intensely on dictionaries would better have a good knowledge of the hash algorithm used for keys. For example, did you know that in plain python there are trivial hash collisions
```
>>> hash((-1,-1))
960947337673586213
>>> hash((-1,-2))
960947337673586213
>>> hash((-2,-1))
960947337673586213
>>> hash((-2,-2))
960947337673586213
```
To me, python hashes are more a convenience than an algorithmic bullet proof feature.

**Comment by nbruin** [comment]
> did you know that in plain python there are trivial hash collisions

Yes, that is due to the fact that "hash" uses -1 to signal an exceptional situation, so -1 hashes to -2. But that's a single collision. We're talking about putting nearly the whole field in one hash bucket.
We know hash functions are not going to be injective. I am aware that it may be the case that fraction field elements may be used as keys, so that there is code out there that relies on them being hashable. I think it should be examined how bad the breakage is if we remove their hashability. That's an easy test: just remove the hash and run the doctests. That at least gives an indication for what kind of breakage occurs. If it's just code that explicitly uses fraction field elements as keys then it's feasible to deprecate: people should just not use them as keys and instead take a numerator-denominator pair. If there are big internal problems (loads of doctests failing that have no obvious hashing involved) then we really need the hash and then we should probably make more of an effort of implementing reasonable ones for common fraction fields.

**For deprecation:** We can just change the hash to almost-constant AND deprecate. Then the error is gone AND we're not falsely advertising something to be hashable.

Note that in most function field extension we can still canonicalize the denominator, just as for number fields: force the denominator to be of higher degree, in the rational function field at the bottom, where forcing the denominator to be monic does the trick.

Silently changing a hash to be nearly constant is really laying a booby trap in the code, so I think we should really make a serious effort to avoid doing so.

**Comment by videlec** [comment]
If there is no agreement I would suggest that we stop this discussion going into circles. It would be nice that you provide an alternative PR to #40000 with your proposal and that we proceed with a vote on sage-devel. I am not spending any more time on this topic which is far away from my priority list.

**Comment by nbruin** [comment]
> Mutable objects are not hashable. But all others should be. It makes sense to design a reasonable hash function so that hash tables are used smartly. But I do not see why this should be mandatory.

Do you have a reference for that one? I see why mutable objects cannot be hashable (mainly: hashable objects can only be allowed to mutate to the degree that their equality properties and hash don't change), but I don't see why immutable objects must be hashable. Is there a design decision in sage that requires it? It may well be that fraction field elements already de facto need to have a hash with them and that changing that is painful ...

**Comment by nbruin** [comment]
Perhaps it's not as bad (in the context of sage) that objects without canonical forms that are equal do not hash to the same value as it would initially seem. It certainly has precedent:
```
sage: a=sin(x)^2
sage: b=1-cos(x)^2
sage: {a:1,b:2}
{sin(x)^2: 1, -cos(x)^2 + 1: 2}
sage: bool(a==b)
True
```
That's a structure in which equality can't even always be determined, so the problem is a bit deeper, but it does show there are "best effort" hashes in sage that miss equalities. It really depends on the application which one is better.  For caching dictionaries (a place where hashes are often used in sage), better performance is probably more important than the occasional missed equality.

So I'm not convinced a nearly-constant hash will improve sage from a practical point of view. I do think deprecating hashability is infeasible because of already the number of doctests that are affected by it (see https://github.com/sagemath/sage/pull/40000#issuecomment-2828958898).

Things that would definitely improve the hashing situation is to properly hash in the many common cases where we can (fix our denominator normalization).

**Comment by nbruin** [comment]
I has a stab at solving the issue more properly at #40019 

Apart from some peculiarities in our inheritance graph, I also found that hashing of fraction field elements happens quite a bit, so there's a real chance that #40000 would be an unacceptable performance loss. I also found that in all the examples I could think of where we can't readily normalize denominators, the "reduce" method already fails (because the gcd computation runs into problems there as well) which led to "hash" producing an error already. So I'm now actually pretty confident that the only code out there using hashes of fraction field elements is for fraction fields in which we DO have a reasonable way of implementing a proper hash.
